test/integration/helpers_test.go

```diff
@@ -0,0 +1,168 @@
+		return nil, err
+	}
+
+	return gw, nil
+}
+
+// gatewayHealthCheck checks the gateway has been scheduled by KIC. This function is inspired by
+// assert.eventually (https://github.com/stretchr/testify/blob/v1.7.5/assert/assertions.go#L1669-L1700)
+// and patched with custom behavior to determine the health of the gateway and to return an error
+// instead of failing (at the time of its call, we don't have any test instance to make fail yet).
+func gatewayHealthCheck(ctx context.Context, client *gatewayclient.Clientset, gatewayName, namespace string) error {
+	ch := make(chan bool, 1)
+
+	timer := time.NewTimer(gatewayWaitTimeToVerifyScheduling)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for tick := ticker.C; ; {
+		select {
+		case <-timer.C:
+			return fmt.Errorf("the gateway scheduling %s/%s has never occurred", namespace, defaultGatewayName)
+		case <-tick:
+			tick = nil
+			go func() {
+				ch <- func() bool {
+					gw, err := client.GatewayV1alpha2().Gateways(namespace).Get(ctx, gatewayName, metav1.GetOptions{})
+					exitOnErr(err)
+					for _, cond := range gw.Status.Conditions {
+						if cond.Reason == string(gatewayv1alpha2.GatewayReasonReady) {
+							return true
+						}
+					}
+					return false
+				}()
+			}()
+		case v := <-ch:
+			if v {
+				return nil
+			}
+			tick = ticker.C
+		}
+	}
+}
+
+// eventuallyGatewayIsLinked checks if the specific Route (HTTP, TCP, TLS, or UDP)
+// is correctly linked to (or unlinked from) a supported gateway. In order to assert
+// that the route must be linked to the gateway, or unlinked from the gateway, the
+// ensureLinked boolean arg must be set accordingly.
+func eventuallyGatewayIsLinked(t *testing.T, c *gatewayclient.Clientset, ensureLinked bool, protocolType gatewayv1alpha2.ProtocolType, namespace, name string) {
+	var routeParents []gatewayv1alpha2.RouteParentStatus
+
+	require.Eventually(t, func() bool {
+		// gather a fresh copy of the route, given the specific protocol type
+		switch protocolType {
+		case gatewayv1alpha2.HTTPProtocolType:
+			route, err := c.GatewayV1alpha2().HTTPRoutes(namespace).Get(ctx, name, metav1.GetOptions{})
+			require.NoError(t, err)
+			routeParents = route.Status.Parents
+		case gatewayv1alpha2.TCPProtocolType:
+			route, err := c.GatewayV1alpha2().TCPRoutes(namespace).Get(ctx, name, metav1.GetOptions{})
+			require.NoError(t, err)
+			routeParents = route.Status.Parents
+		case gatewayv1alpha2.UDPProtocolType:
+			route, err := c.GatewayV1alpha2().UDPRoutes(namespace).Get(ctx, name, metav1.GetOptions{})
+			require.NoError(t, err)
+			routeParents = route.Status.Parents
+		case gatewayv1alpha2.TLSProtocolType:
+			route, err := c.GatewayV1alpha2().TLSRoutes(namespace).Get(ctx, name, metav1.GetOptions{})
+			require.NoError(t, err)
+			routeParents = route.Status.Parents
+		default:
+			t.Fatalf("protocol %s not supported", string(protocolType))
+		}
```

Comment:
I agree with you that this approach can be better to improve code clarity and usability. I'll follow your suggestion.

Reply:
Can you post here @mlavacca how do you see the debugging being hindered?

To me, the retrievals of the gateway - via `Get()` - should also be retried, shouldn't they? Not fail the test outright?

Ideally we would log (using e.g. `t.Log()`) from `Eventually()` and similar and return false instead of using `require` and/or `assert` since that marks the test as failed

Here's an issue I create for using `require` and/or `assert` in `Eventually()` and other similar assertion helpers: https://github.com/Kong/kubernetes-ingress-controller/issues/2629

Reply:
> To me, the retrievals of the gateway - via Get() - should also be retried, shouldn't they? Not fail the test outright?

Ah yeah, this gets into some deeper bits about how our tests are written. I think one could definitely make this argument in general given the eventually consistent nature of the API we are interfacing with, however more practically speaking in a testing environment it is often the case that we have a clear and pointed order of operations that we expect to be acted upon in a reproducible manner.

So while good controller code ultimately is re-entrant and retries operations, I think an argument could be made that test code can and maybe should be more linear and rigid. Consider the following example:

```golang
require.NoError(t, k8s.Create(ctx, obj))
require.Eventually(t, func() bool {
    obj, err = k8s.Get(ctx, obj.Name)
    require.NoError(t, err)
    return obj.Status.Ready == true
}, timeout, tick)
```

The fact that the `Create` did not return an error should incontrovertibly mean that the following `Get` calls should not fail _unless_ there is some kind of critical failure in the testing environment, right? But at the same time, it's also reasonable to retry since we're talking about asynchronous processes going on the effects of which we're testing for right? Both are right in their own way, and both are wrong in there own way.

Ultimately I think it's slight weakness of our current testing code that with `Get()` loops we're checking _that_ something happens but we have limited testing for _when_ it happens. In a controlled testing environment the sequence of events for an object is most often going to be predictable and reproducible so we could probably make this better by adding some convenient functionality to KTF that uses something like informer semantics to track the progress and sequence of changes to an object. That's not something that I feel we need to burden this PR with, so I've created https://github.com/Kong/kubernetes-testing-framework/issues/315 for that.

In the meantime, I think it's reasonable to not have retries for actions that are expected to succeed and we have no quantification of any situations where they wouldn't (that aren't otherwise some kind of significant problem with the testing environment). Simply returning errors quickly in these situations would seem to likely actually catch more problems that may have otherwise gone missed.

Reply:
The drawback of that approach is that we might miss an intermediate state, that switched too fast, and fail the test even thought the end state is correct.

Reply:
> The drawback of that approach is that we might miss an intermediate state, that switched too fast, and fail the test even thought the end state is correct.

Yes, that's exactly what I was trying to talk about above and is why I created https://github.com/Kong/kubernetes-testing-framework/issues/315 as it would enable catching the intermediate states. I don't feel we need to block this PR on these kinds of things though, as this problem existed before this PR and it would significantly slow down this PR to tack on an extra project within it. Better to move quickly with small iterations and keep track of technical debt like this so we can fix it forward. My suggestion is that we accept this as-is to keep this PR from becoming swamped and follow up on more fundamental testing issues as their own iteration and tasks.

Reply:
I consider this comment resolved, we've spent enough time on it we don't need to spend more, and as you've correctly identified this is a common problem we need to approach holistically and not burden an unrelated change with.